---
title: "shrunken eQTL-tissue visualization "
author: "Kushal Kumar Dey"
date: "Friday, June 05, 2015"
output: html_document
---

```{r packages}
suppressMessages(suppressWarnings(library(tsne)))
suppressMessages(suppressWarnings(library(Rtsne)))
suppressMessages(suppressWarnings(library(Seurat)))
library(qtlcharts)
#library(iplots)

```


##  Strongest eQTLs first batch

```{r data_read,eval=FALSE}
setwd(list.dirs("/Users/kushal/Documents/Matthew Stephens Project/gtex-viz-kush/",recursive=FALSE)[2])
strongmeans_eqtl=read.table("../data/max_tscore_eQTL.table.txt");
strongmeans_eqtl_complete=strongmeans_eqtl[complete.cases(strongmeans_eqtl),-c(1,47)];
# write.table(strongmeans_eqtl_complete,"../data/complete_tscore_eQTL.table.txt")
# write.table(strongmeans_eqtl_complete[,1],"../data/gene_snp_names.txt")
```

Normalize effects so largest effect always positive and (optionally) divide by L2 norm to make effects "comparable".

```{r norm_trans, eval=FALSE}
L2norm = function(x){return(sqrt(sum(x^2)))}

norm_effects = function(m,standardize=TRUE){
  max_pos = apply(m,1,max)
  max_neg = apply(-m,1,max)
  max_sign = ifelse(max_pos>max_neg, 1,-1) #find the sign of the largest effect
  if(standardize){max_sign=max_sign* apply(m,1,L2norm)}
  m/max_sign
}
strongmeans_eqtl_norm= norm_effects(strongmeans_eqtl_complete[,-1]);
write.table(strongmeans_eqtl_norm,"../data/normalized_tscore_eQTL.table.txt")


```

```{r tissue_names}
names_tissues <-  read.table("../data/names_tissues.txt")[-1];

tissue_names_filtered=unlist(lapply(as.matrix(names_tissues),
                            function(x) substring(x,32,nchar(x)-7)));
```

##  SFA on the eQTLs

We perform SFA on this first batch of eQTLs to see if any clustring pattern shows up.

```{r sfa_eqtl}

lambda=as.matrix(read.table("../data/sfa_MatrixEqtl_15/sfa_MatrixEqtl_15_lambda.out"));
fac_mat=as.matrix(read.table("../data/sfa_MatrixEqtl_15/sfa_MatrixEqtl_15_F.out"));

K=15;
color=c("red","blue","cornflowerblue","black","cyan","darkblue",
  "brown4","burlywood","darkgoldenrod1","darkgray","deepskyblue","darkkhaki",
  "firebrick","darkorchid","hotpink","green","magenta","yellow", "azure1","azure4");

lambda_pos=apply(lambda,c(1,2),function(x) max(x,0));
lambda_neg=apply(lambda,c(1,2),function(x) min(x,0));

for(i in 1:dim(lambda)[2]){
Sys.sleep(0.1)
b=barplot(fac_mat[i,],las=2,ylim=c(-8,8),xaxt="n")
#text(b, fac_mat[i,],labels=round(fac_mat[i,],2), pos=3, offset=.05)
axis(1,at=seq(0.5,52.5,length.out=44),tissue_names_filtered,las=2)
title(paste0("Patterns of effect for factor",i));
}


```


The t-SNE on eQTLs in 2-D and 5-D is performed

```{r cache=TRUE}
tsne_eqtl_shrunk_2 <- tsne(strongmeans_eqtl_first_norm,k=2);
tsne_eqtl_shrunk_5 <- tsne(strongmeans_eqtl_first_norm,k=5);
```




We visualize the tSNE on 2-D as follows

```{r tsne_eqtl_2}
#suppressMessages(suppressWarnings(iplot(tsne_eqtl_shrunk_2[,1],tsne_eqtl_shrunk_2[,2])))




gene_names_eqtl_study=substring(strongmeans_eqtl_first[,1],1,15);

write.table(gene_names_eqtl_study,"D:/Matthew_Stephens_Project/gTEx/gtex-viz-master/data/gene_names_eqtl_study.txt");

gene_info_eqtl_study=read.table("D:/Matthew_Stephens_Project/gTEx/gtex-viz-master/data/gene_info_eqtl_study.txt");

matching_labels_eqtls_to_gene_info= match(gene_info_eqtl_study[,1],gene_names_eqtl_study);

tsne_eqtl_shrunk_2_matched=tsne_eqtl_shrunk_2[matching_labels_eqtls_to_gene_info,];

temp=tsne_eqtl_shrunk_2_matched;

rownames(temp)=paste0(strongmeans_eqtl_first[matching_labels_eqtls_to_gene_info,1],", ",
                      "name-",gene_info_eqtl_study$external_gene_name,", ",
                      "descr-",gene_info_eqtl_study$description,",",
                      "chr-",gene_info_eqtl_study$chromosome_name, ", ",
                      "gc-",gene_info_eqtl_study$percentage_gc_content);


strongmeans_eqtl_first_norm_matched=strongmeans_eqtl_first_norm[matching_labels_eqtls_to_gene_info,];
rownames(strongmeans_eqtl_first_norm_matched)=rownames(temp);
colnames(strongmeans_eqtl_first_norm_matched)=paste0("V_",1:44)


suppressMessages(suppressWarnings(iplotCurves(strongmeans_eqtl_first_norm_matched, 1:44, temp)));

```

We want to use the gene names to extract the columns in the reads dataset. In order to accomplish this, we extract just the gene names, removing the SNP name.

```{r gene_names_eval,echo=FALSE, eval=FALSE}


temp=gsub("^.*?_","_","ATGAS_1121")

temp_str=gsub("_","\t",as.character(strongmeans_eqtl_first[,1]));
gene_names_w_frac=unlist(lapply(temp_str, function(x) unlist(strsplit(x,"\t"))[1]));

write.table(gene_names_w_frac, "D:/Matthew_Stephens_Project/gTEx/gtex-viz-master/data/gene_names_w_frac.txt");


lapply(strongmeans_eqtl_first[,1],function(x) strsplit(gsub("_","$",x),'$'))

```




```{r qtlplot_nontransform}

temp1=temp;
strongmeans_eqtl_first_mat_matched=strongmeans_eqtl_first_mat[matching_labels_eqtls_to_gene_info,];
rownames(strongmeans_eqtl_first_mat_matched)=rownames(temp1);
colnames(strongmeans_eqtl_first_mat_matched)=paste0("V_",1:44)

suppressMessages(suppressWarnings(iplotCurves(strongmeans_eqtl_first_mat_matched, 1:44, temp1)));


```

We also present the gene level reads expression forthe cis-genes in these cis-eQTLs. However a better representation is made in a separate file (*vis_eqtl_cis_gene_reads.R* since there seems to be a bug in qtlcharts regarding having multiple iplotCurves in same file, which prevents 2-way mapping from the second iplotCurve onwards and we can map in only one direction).


```{r cis_gene_reads}

reads_cis_genes= read.table("D:/Matthew_Stephens_Project/gTEx/gtex-viz-master/data/matching_reads_with_genes_in_file.txt");

sample_id <-  read.table("D:/Matthew_Stephens_Project/gTEx/gtex-viz-master/data/samples_id.txt")[,3];

reads_cis_gene_tissue=t(apply(reads_cis_genes,1,function(x) tapply(x,sample_id,mean)));

temp2=temp;
reads_cis_gene_tissue_matched=reads_cis_gene_tissue[matching_labels_eqtls_to_gene_info,];
rownames(reads_cis_gene_tissue_matched)=rownames(temp2);
colnames(reads_cis_gene_tissue_matched)=paste0("V_",1:53)

suppressMessages(suppressWarnings(iplotCurves(log(reads_cis_gene_tissue_matched+1), 1:53, temp2)));

```







Next, we visualize the t-SNE in 5-D using projections 1,2 and 3,4 and 4,5.

```{r tsne_eqtl_5}
suppressMessages(suppressWarnings(iplot(tsne_eqtl_shrunk_5[,1],tsne_eqtl_shrunk_5[,2])))

suppressMessages(suppressWarnings(iplot(tsne_eqtl_shrunk_5[,3],tsne_eqtl_shrunk_5[,4])))


suppressMessages(suppressWarnings(iplot(tsne_eqtl_shrunk_5[,4],tsne_eqtl_shrunk_5[,5])))

```

We also look at the t-SNE plot of the samples based on this first batch of eQTL data as follows (although going by the previous analysis, we are not very hopeful that it would show us some nice patterns)



```{r cache=TRUE}

tsne_samples <- tsne(t(strongmeans_eqtl_first_norm));
```

```{r tsne_sample_viz}
plot(tsne_samples,type="n")
text(tsne_samples[,1],tsne_samples[,2],tissue_names_filtered,cex=0.5)

```

